### PR TITLE
Make the --dynamic flag create a shared library when CHPL_LLVM=llvm

### LINKED ISSUE #12699: The --dynamic flag should create a shared library when compiling with --llvm
Currently, it looks like the LLVM backend creates a static library irregardless of whether `--static` or `--dynamic` is passed to Chapel as an argument.

The `makeBinaryLLVM()` function in `compiler/llvm/clangUtil.cpp` branches to `makeLLVMLibrary()` when compiling a library. This function creates a static library using the `ar` utility.

It appears that no additional linking step is taken in the case of `--dynamic`. If this is indeed the case (and I'm not missing something), it would be a simple fix to invoke the linker with the correct arguments to create a shared library.

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Previously, Chapel would create a static library using the `ar` utility even when the `--dynamic` flag was set. This PR repurposes an existing invocation of `clang++` (previously used when linking binaries) to properly link dynamic libraries.

For best results, compile with `CHPL_LIB_PIC=pic`, as many linkers (GNU gold, for example) will refuse to link dynamic libraries when there are text relocations present.

Conditional code is needed to keep `dyld` from failing to load a shared library when on `darwin`. When the shared library is first produced, the internal `install_name` stored in the library file is the path to the temporary copy of the library (which is eventually removed). This conditional code changes the `install_name` so that it matches the library's final output location.

This code is only relevant when using the default linker on `darwin`. Not all linkers may have the `install_name` flag, so it would be best to _make this code depend on the linker being used_ in addition to being on `darwin`. I have not done this, and am not sure how. It would require fetching the linker being used directly from `clang++`. 

Testing:

- [x] `ALL` on `linux64` when `CHPL_LLVM=llvm`
- [x] `interop/C/llvm/makefiles` on `darwin` when `CHPL_LLVM=llvm`

Resolves #12699.

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Should I add the test to the `test/interop/C/llvm/makefiles` directory, or should I make a new directory for this in `test/interop/C/llvm`? (Or maybe, there is a better place to put the test)?

I was thinking that some additional tests to validate dynamic linking would be good.

**Comment by lydia-duncan** [comment]
I don't think a better place exists.  Making your own subfolder for dynamic seems like it would make the test the most visible, but I wouldn't be against having it just be a side test in another directory as long as it was clearly labeled

**Comment by dlongnecke-cray** [comment]
I ended up putting the test in `interop/C/llvm/makefiles/dynamicLinking`, let me know if you'd like me to move it. Running another full test right now.

File: test/interop/C/llvm/makefiles/dynamicLinking/FooLibrary.chpl [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Hmmmmmmmmmm...
```
Comment (lydia-duncan) [inline test/interop/C/llvm/makefiles/dynamicLinking/FooLibrary.chpl]: A slightly better opening comment would be good XD

File: test/interop/C/llvm/makefiles/dynamicLinking/useSharedLibFromC.test.c [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Make sure we link appropriately from C.
+#include "lib/FooLibrary.h"
+
+extern void wisdom(void);
```
Comment (lydia-duncan) [inline test/interop/C/llvm/makefiles/dynamicLinking/useSharedLibFromC.test.c]: This shouldn't be needed, the .h file should handle it for you

**Comment by dlongnecke-cray** [comment]
Just one more update before I merge, all tests pass (except for gsl_demo...) after the latest changes.

**Pull request merged**